### DISCUSSION TITLE
Question on best practice for Supabase branching for development and production

### DISCUSSION THREAD

**Discussion by wusixuan0** [supabase/supabase#27860] [Questions] [answered]
Hi guys! I'm seeking clarification on the best practice for implementing Supabase branching in my project. The documentation warns against using branching in a "production setting," but also describes merging branches within a single project. I'm unsure whether I should create two separate Supabase projects (one for production and one for development with branches) or use a single project with a main "production" branch and separate development branches. My app is still in the development stage, but I want to set up a proper workflow that separates development and production environments while leveraging the new branching feature. How should I structure my Supabase project(s) to best utilize branching while following best practices for development and eventual production use? Thank you so much for your help!

**Comment by wusixuan0** [comment]
I think I found the correct answer from Supabase docs and Supabase Chat: visit [supabase.com/docs](https://supabase.com/docs) and hit cmd + /
- docs warns against using the Branching feature in a production setting, and recommends using Migrations to manage changes to your production database.
- it is recommended to have separate projects for development and production environments in Supabase. 
- Branching is an early access feature for test and preview changes.

**Reply by sweatybridge** [reply, marked as answer]
Hi @wusixuan0, we have updated our [branching docs](https://supabase.com/docs/guides/deployment/branching) recently to clarify the best practices for development and promoting to production. Typically we would recommend

- A single production project with branching enabled.
- Each pull request creates a development branch.
- Add all [database migrations](https://supabase.com/docs/guides/deployment/database-migrations) through PRs.
- When you merge a PR, branching will automatically apply changes from your migration files to your production project.

Feel free to reach out if you have other questions.

**Reply by code-leen** [reply]
@sweatybridge Hello! How would we go about this if our production and test environment databases are in separate Supabase projects AND use the same GitHub repo? Is it still possible to use the branching feature?

**Reply by sweatybridge** [reply]
> @sweatybridge Hello! How would we go about this if our production and test environment databases are in separate Supabase projects AND use the same GitHub repo? Is it still possible to use the branching feature?

You can setup a persistent branch to be your staging environment. It behaves the same as a separate project, with the benefit that when you merge a PR from staging git branch to prod, your database changes will also be automatically promoted to prod.

**Reply by code-leen** [reply]
@sweatybridge To clarify, we could replace our test Supabase project with a persistent branch (as our staging/test environment) on our production Supabase project? Thank you so much for the quick reply!!

**Reply by sweatybridge** [reply]
Yup! That's exactly right.

To migrate from your staging project to a persistent branch, you will need to first create a persistent branch, then perform a [backup and restore](https://supabase.com/docs/guides/platform/migrating-within-supabase/backup-restore) from your staging project to the branch.

**Reply by code-leen** [reply]
@sweatybridge That's amazing! Thank you for sharing!!
